src/analysis/lattices/array.h

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2023 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_analysis_lattices_array_h
+#define wasm_analysis_lattices_array_h
+
+#include <array>
+#include <utility>
+
+#include "../lattice.h"
+#include "bool.h"
+#include "flat.h"
+
+namespace wasm::analysis {
+
+// A lattice whose elements are N-tuples of elements of L. Also written as L^N.
+template<Lattice L, size_t N> struct Array {
+  using Element = std::array<typename L::Element, N>;
+
+  L lattice;
+
+  Array(L&& lattice) : lattice(std::move(lattice)) {}
+
+private:
+  // Use a template parameter pack to generate N copies of
+  // `lattice.getBottom()`. TODO: Use C++20 lambda template parameters instead
+  // of a separate helper function.
+  template<size_t... I>
+  Element getBottomImpl(std::index_sequence<I...>) const noexcept {
+    return {((void)I, lattice.getBottom())...};
+  }
+  template<size_t... I>
+  Element getTopImpl(std::index_sequence<I...>) const noexcept {
+    return {((void)I, lattice.getTop())...};
+  }
+
+public:
+  Element getBottom() const noexcept {
+    return getBottomImpl(std::make_index_sequence<N>());
+  }
+
+  Element getTop() const noexcept
+#if __cplusplus >= 202002L
+    requires FullLattice<L>
+#endif
+  {
+    return getTopImpl(std::make_index_sequence<N>());
+  }
+
+  // `a` <= `b` if their elements are pairwise <=, etc.
```

Comment:
Maybe stress that this is not the same as dictionary sorting (which would stop at the first difference)?